ui/app/AppLayouts/Wallet/popups/buy/BuyCryptoModal.qml

```diff
@@ -17,78 +17,153 @@ import utils 1.0
+                d.urlIsBeingFetched = false
+                if (!!d.selectedProviderEntry.item && !!url)
+                    Global.openLinkWithConfirmation(url, d.selectedProviderEntry.item.hostname)
+                root.close()
+            }
+        }
+
+        // used to filter items based on search string in the token selector
+        property string searchString
+        readonly property var tokenSelectorViewAdaptor: TokenSelectorViewAdaptor {
+            assetsModel: root.groupedAccountAssetsModel
+            plainTokensBySymbolModel: root.plainTokensBySymbolModel
+            flatNetworksModel: root.networksModel
+            currentCurrency: root.currentCurrency
+
+            showAllTokens: true
+            enabledChainIds: root.buyCryptoInputParamsForm.selectedNetworkChainId !== -1 ? [root.buyCryptoInputParamsForm.selectedNetworkChainId] : []
+            accountAddress: root.buyCryptoInputParamsForm.selectedWalletAddress
+            searchString: d.searchString
+        }
+
+        readonly property var buyCryptoAdaptor: BuyCryptoModalAdaptor {
+            networksModel: root.networksModel
+            areTestNetworksEnabled: root.areTestNetworksEnabled
+            processedTokenSelectorAssetsModel: d.tokenSelectorViewAdaptor.outputAssetsModel
+            selectedProviderSupportedAssetsArray: {
+                if (!!d.selectedProviderEntry.item && !!d.selectedProviderEntry.item.supportedAssets)
+                    return ModelUtils.modelToFlatArray(d.selectedProviderEntry.item.supportedAssets, "key")
+                return null
+            }
+            selectedChainId: root.buyCryptoInputParamsForm.selectedNetworkChainId
+        }
     }
 
     width: 560
     height: 515
     padding: Style.current.xlPadding
-    stackTitle: qsTr("Buy assets for %1").arg(!!buyCryptoAdaptor.selectedAccount ? buyCryptoAdaptor.selectedAccount.name: "")
+    stackTitle: qsTr("Buy assets for %1").arg(!!d.selectedAccountEntry.item ? d.selectedAccountEntry.item.name: "")
     rightButtons: [d.buyButton, finishButton]
     finishButton: StatusButton {
         text: qsTr("Done")
         onClicked: root.close()
     }
 
-    Connections {
-        target: root.buyCryptoAdaptor
-        function onProviderUrlReady(url) {
-            if (!!root.buyCryptoAdaptor.selectedProvider && !!url)
-                Global.openLinkWithConfirmation(url, root.buyCryptoAdaptor.selectedProvider.hostname)
-            root.close()
-        }
-    }
-
-    onOpened: root.buyCryptoAdaptor.fetchProviders()
+    onOpened: root.buyCryptoStore.fetchProviders()
     onClosed: {
         // reset the view
+        d.uuid = ""
+        d.urlIsBeingFetched = false
         root.replaceItem = undefined
         buyCryptoProvidersListPanel.currentTabIndex = 0
-        root.buyCryptoAdaptor.reset()
         root.buyCryptoInputParamsForm.resetFormData()
     }
+    Component.onCompleted: {
+        root.buyCryptoStore.providerUrlReady.connect(d.providerUrlReady)
```

Comment:
I'm wondering if we're not missing an error state. Probably would be nice to inform the user and allow a retry.

Reply:
good point, we discussed to add error handling in 2.31 so we've not handled it here yet.

Reply:
I can add a TODO